test/token/ERC20/extensions/ERC4626.test.js

```diff
@@ -21,6 +22,30 @@ contract('ERC4626', function (accounts) {
     }
   });
 
+  it('asset has not yet been created', async function () {
+    for (let i = 0; i < 6; i++) {
+      const vault = await ERC4626.new('', '', accounts[i]);
+      expect(await vault.decimals()).to.be.bignumber.equal(decimals);
+    }
+  });
+
+  it('underlying excess decimals', async function () {
+    const token = await ERC20ExcessDecimalsMock.new();
+    const vault = await ERC4626.new('', '', token.address);
+    expect(await vault.decimals()).to.be.bignumber.equal(decimals);
+  });
+
+  it('decimals overflow', async function () {
+    for (const offset of [243, 250, 255].map(web3.utils.toBN)) {
+      const token = await ERC20Decimals.new('', '', decimals);
+      const vault = await ERC4626OffsetMock.new(name + ' Vault', symbol + 'V', token.address, offset);
+      await expectRevert(
+        vault.decimals(),
```

Comment:
In that case, we should probably revert in the constructor.

Reply:
Is it really worth the additional deployment costs?

Reply:
I'd argue that if it fails because of a `decimals()` overflow, it's behaving as expected (that's why we're adding the test). I stand with @pcaversaccio and don't think it should go on the constructor

Reply:
Also, we are talking about a valut where is offset is so big that `10**offset` doesn't fix in a uint256. This means `_convertToShares` and `_convertToAssets` will always revert. That means this contract is completely unusable.

I'm thinking there should be an upper bound on the offset. Anything beyond 64 is really dangerous. I personally don't think it should ever be more than 32

Reply:
Well, the functions we're talking about here (`decimals`, `_convertToShares`, and `_convertToAssets`) are all `virtual`. So let's say we add a check in the constructor. Someone inherits from the `ERC4626` abstract contract and implements a new `_convertToShares` and `_convertToAssets` function (for whatever reason) without a `super` call to the parent function. In these new functions anything can be possible, e.g. `1**offset` or `100**offset`, which could lead to a non-overflowing or overflowing behaviour even with having a check. If we're talking about that high of an offset, I think the problem itself is a PEBCAK problem.

Reply:
Lets not get confused here:

The fact that devs might override in way that are unsafe is beyond our control, and the scope of what we can check.

What I believe we should be concerned about is the fact that we currently have a contract that, without modification from the dev, can be completely broken if the constructor parameters are not within a certain range.

Right now, all it takes is a deployer to not fully understand the way the decimal offset work, use a big value, and end up with a bricked contract. I don't think this is OK. I believe its better to revert during the deployment than to end up with a contract where every meaningful operation will revert.

This is already something we do in the `VestingWallet` contract. If the `beneficiaryAddress` is 0, we revert during construction.

Reply:
> What I'm concerned of is the corresponding revert string to get written on the bytecode if we set an upper bound. Do you think it's worth it?

You are more concerned with a revert string in th deployment bytecode (it shouldn't even make it into the actual bytecode of the contract) than with broken contracts being deployed ?

Reply:
>  Right now, all it takes is a deployer to not fully understand the way the decimal offset work, use a big value, and end up with a bricked contract.

First of all, I do fully agree that this is not ok - that's why I added this test in the first line. But, just for me to understand this argument: someone who is able to understand what a `virtual` function is and who is able to override the default `_decimalsOffset` in a way that breaks the contract does not understand how the offset works? What if we document this properly in the function `_decimalsOffset`, because this is what people will read first in case they override it?

> What I believe we should be concerned about is the fact that we currently have a contract that, without any modification from the dev, can be completely broken if the constructor parameters are not within a certain range.

I don't think this is true. The testing mock is configured in a customisable way, yes, but the `ERC4626` itself is an abstract contract that does not have an `offset` value set in the constructor (the max value for `decimals` in the default setting is already constrained to `type(uint8).max`). It's set by default to 0. Thus, what I want to emphasise here, is that the `ERC4626` (abstract) contract is not vulnerable by default to the overflow. What it takes is to customise the offset itself. Please correct me if I'm wrong here.

Reply:
> but the `ERC4626` itself is an abstract contract that does not have an `offset` value set in the constructor

Yes indeed. My mistake. I confused the mock constructor with the contract constructor.

Reply:
I was confused by the fact that at some point `offset` was an immutable variable set in ERC4626 constructor. All good !